Add program logger

alexdewar opened the pull request:
# Description

For various reasons, it's better to use a proper logging library rather than having print statements dotted throughout the code.

I've added `env_logger` as a dependency (it seems to be a conservative choice) and used it in a few places in the code. The default output will look something like this:

```
[2024-07-02T15:21:11Z INFO  muse2::simulation] Loading settings from "examples/simple/settings.toml"
```

We can always configure it later if we want it to look different.

Closes #91.

## Type of change

- [ ] Bug fix (non-breaking change to fix an issue)
- [x] New feature (non-breaking change to add functionality)
- [ ] Refactoring (non-breaking, non-functional change to improve maintainability)
- [ ] Optimization (non-breaking change to speed up the code)
- [ ] Breaking change (whatever its nature)
- [ ] Documentation (improve or add documentation)

## Key checklist

- [x] All tests pass: `$ cargo test`
- [x] The documentation builds and looks OK: `$ cargo doc`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

dalonsoa commented on src/main.rs:
```diff
@@ -10,10 +10,20 @@ use std::path::Path;
 /// The main entry point to the program
 fn main() {
+    // Log at the info level by default, unless the user has specified otherwise
+    if env::var("RUST_LOG").is_err() {
+        env::set_var("RUST_LOG", "info")
```
I'm not sure controlling the log level via environment variables - specially such general ones like one called `RUST_LOG` that will affect any rust tool using the same logging tool - is such a user friendly idea. Also, there is a more elegant way of setting the default level, if not provided: https://docs.rs/env_logger/latest/env_logger/#specifying-defaults-for-environment-variables

At the very least, I would allow the user to configure the log level via input arguments or the settings file. IN MUSEv1 this is done like:

```python
        settings = read_settings(settings)
        getLogger("muse").setLevel(settings.log_level)
        MCA.factory(settings).run()
```

dalonsoa reviewed (changes requested):
Having a logger is important, but we need to make it more user friendly, I think.

alexdewar commented:
@dalonsoa I've had another go, addressing your comments.

Now users can either set the logging level via an environment variable (`MUSE2_LOG_LEVEL`) or the `settings.toml` file, with the environment variable taking precedence (I figure it's probably handy to have this for debugging). As we want to be able to use the program logger to log progress as we load the various input files, we need to initialise the logger after reading `settings.toml` (so we know if the user has configured the log level or not), but before loading any of the CSV files. One option would have been to initialise the logger in the `read_settings()` function, but that seemed like a poor separation of concerns. Instead, I refactored the code so that reading the TOML file and reading the other files are carried out in two steps; in between them, we can extract the log level and set it. I've refactored `settings.rs` using a more idiomatic Rust style. Now the struct with the raw input data from `settings.toml` is called `SettingsReader`. It provides a method for accessing the log level and for converting it into a `Settings` struct, which is the higher-level representation of the model settings. I also followed a suggestion you made somewhere else about reading the settings in `main()`, then passing them as an argument to `simulation::run()`, which I think is cleaner.

dalonsoa reviewed (approved):
Looks good! But it seems there are quite a few things not covered by tests...

alexdewar commented:
Yeah :disappointed:. Part of the issue is that mocking in Rust is really hard, but also I've just been a bit lazy :stuck_out_tongue:. Some of the bits that are untested were untested before though, it's just that codecov is flagging them again because of the refactoring.

Honestly I'm not sure what the best approach is for testing in Rust. For little functions, it's easy enough to write a unit test, but for higher-level functions that call these little functions, it's hard to think of a sensible test. I guess maybe we just write functional tests for these ones?

dalonsoa commented:
Yeah, I think `rust` encourages write code in small chunks that are easy to test. Anything bigger than that, looks more like an integration test.
